sweep/store.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+)
+
+var (
+	// lastTxBucketKey is the key that points to a bucket containing a
+	// single item storing the last published tx.
+	//
+	// maps: lastTxKey -> serialized_tx
+	lastTxBucketKey = []byte("sweeper-last-tx")
+
+	// lastTxKey is the fixed key under which the serialized tx is stored.
+	lastTxKey = []byte("last-tx")
+
+	// txHashesBucketKey is the key that points to a bucket containing the
+	// hashes of all sweep txes that were published successfully.
+	//
+	// maps: txHash -> empty slice
+	txHashesBucketKey = []byte("sweeper-tx-hashes")
+
+	// utxnChainPrefix is the bucket prefix for nursery buckets.
+	utxnChainPrefix = []byte("utxn")
+
+	// utxnHeightIndexKey is the sub bucket where the nursery stores the
+	// height index.
+	utxnHeightIndexKey = []byte("height-index")
+
+	// utxnFinalizedKndrTxnKey is a static key that can be used to locate
+	// the nursery finalized kindergarten sweep txn.
+	utxnFinalizedKndrTxnKey = []byte("finalized-kndr-txn")
+)
+
+// SweeperStore stores published txes.
+type SweeperStore interface {
+	// IsOurTx determines whether a tx is published by us, based on its
+	// hash.
+	IsOurTx(hash chainhash.Hash) (bool, error)
+
+	// NotifyPublishTx signals that we are about to publish a tx.
+	NotifyPublishTx(*wire.MsgTx) error
+
+	// GetLastPublishedTx returns the last tx that we called NotifyPublishTx
+	// for.
+	GetLastPublishedTx() (*wire.MsgTx, error)
+}
+
+type sweeperStore struct {
+	db *channeldb.DB
+}
+
+// NewSweeperStore returns a new store instance.
+func NewSweeperStore(db *channeldb.DB) (SweeperStore, error) {
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(
+			lastTxBucketKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		if tx.Bucket(txHashesBucketKey) != nil {
+			return nil
+		}
+
+		txHashesBucket, err := tx.CreateBucket(txHashesBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Use non-existence of tx hashes bucket as a signal to migrate
+		// nursery finalized txes.
+		migrateTxHashes(tx, txHashesBucket)
+
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sweeperStore{
+		db: db,
+	}, nil
+}
+
+// migrateTxHashes migrates nursery finalized txes to the tx hashes bucket. This
+// is not implemented as a database migration, to keep the downgrade path open.
+func migrateTxHashes(tx *bbolt.Tx, txHashesBucket *bbolt.Bucket) error {
+
+	// Iterate over all nursery chain buckets.
+	return tx.ForEach(func(name []byte, chainBucket *bbolt.Bucket) error {
+		// Skip non-nursery buckets
+		if len(name) < len(utxnChainPrefix) {
+			return nil
+		}
+
+		if !bytes.Equal(name[:len(utxnChainPrefix)], utxnChainPrefix) {
+			return nil
+		}
+
+		// Retrieve the existing height index.
+		hghtIndex := chainBucket.Bucket(utxnHeightIndexKey)
+		if hghtIndex == nil {
+			return nil
+		}
+
+		// Retrieve all heights.
+		hghtIndex.ForEach(func(k, v []byte) error {
```

Comment:
Ignoring the error here and a few places below in this method. However the next diff (IMO) should move this into the nursery so we don't need to spill all the information of its bucket structure over.

Reply:
Error handled. Just saw it one more times, fixed that too. Info spill, see comments above.